finalfrontier/src/vocab.rs

```diff
@@ -213,10 +213,29 @@ impl From<SubwordVocab> for VocabWrap {
         .into()
     }
 }
+/// Index representation.
+///
+/// This enum represents possible types of indices, e.g. in a vocabulary with subwords, the `Multi`
+/// variant includes all indices associated with a key.
+#[derive(Clone)]
+pub enum Idx {
+    Single(usize),
+    Multi(Vec<u64>),
+}
+
+impl Idx {
+    /// Return the unique index for the index.
+    pub fn single_idx(&self) -> usize {
+        match self {
+            Idx::Single(idx) => *idx,
+            Idx::Multi(v) => *v.last().unwrap() as usize,
```

Comment:
Now this assumes that the word index is always the last index? This seems quite brittle. I know that this has the convenience of being able to provide a `&[u64]` in various places, but it would be nice to have stronger invariants.

Reply:
Could probably made a tuple `(Vec<usize>, usize)` to get the single focus idx.

Reply:
I have looked through all PRs again, and I think that this is the clearest iteration, though it was still very useful to see the trait version.

I think this is ready to go with two modifications:

1. Changing `Multi` in a tuple as you suggested.
2. Renaming of `Idx` to `WordIdx`, with the methods `word_idx` (applicable to both enum variants) and `subword_indices` (only applicable to `Multi`). I would also consider changing `Single` to `Word` and `Multi` to `WordWithSubwords` (or something along those lines).

The variant renaming would also make it possible to have a `Subwords` variant later with only subwords.

Reply:
Thanks for the feedback! I'm not really happy with any of the 2 (and a half) iterations I proposed so far. I just pushed the cleaned up version anyways.

> The variant renaming would also make it possible to have a `Subwords` variant later with only subwords.

Although implementing `word_idx` on such a variant would be odd.